### ISSUE TITLE
How to persist optional parameter or state data on browser back?

### ISSUE THREAD

**Issue by pankajparkar** [angular-ui/ui-router#2115]
I'm having one parent state that has two children's state inside that I'm going to show one state based on the `URL`.

Out of those two `states` one is having to parameters like `param1` and `param2`, I have use `params` option of ui-router inside state definition.

State

```
$stateProvider.state('tabs.account', {
    url: '/account',
    views: {
        'content@tabs': {
            templateUrl: 'account.html',
            controller: function($scope, $stateParams) {
                //This params are internally used to make ajax and show some data.
                $scope.param1 = $stateParams.param1;
                $scope.param2 = $stateParams.param2;
            },
        }
    },
    params: {
        param1: { value: null }, //this are optional param
        param2: { value: null } //because they are not used in url
    }
});
```

If you look at my route the params option is not really introduced inside the `URL`, that's why I'm considering then as optional. 
## Problem

Look at plunkr, I've shown two tabs **Account** & **Survey**, 
1. Click on Survey tab, then add some data in the `textarea` which are shown. 
2. Click on **Go to Account** that will pass those `textarea` values to
   the other **Account** tab by doing `ui-sref="tabs.account({param1: thing1, param2: thing2})"` on the anchor
3. Now you will see the `param1` & `param2` values on html which has been assigned to scope from `$stateParams`
4. Now again Click on **Survey** tab, you will land on the survey page.
5. Just click browser back, you will notice that `param` value is not getting `null`.

[**Problem Plunkr**](http://plnkr.co/edit/NF0W2apwlz7ufGSu2OCR?p=preview)

I believe you got what I wanted to ask, why the optional parameter value has not been store? as they have been a part of state.

I know I can solve this issue by below two solutions.
- By creating one service that will share data between two views.
- By adding parameter inside the state URL. like `url: '/account/:param1/:param2',` (But i wouldn't prefer this)

I already tried angular-ui-routers [**`sticky states`**](http://christopherthielen.github.io/ui-router-extras/#/sticky) but that doesn't seems to work for me. What is the better way to this?

Is there any technique that I can use for my case, Any ideas would appreciate.

[**Stackoverflow Question here**](http://stackoverflow.com/questions/31542972/how-to-persist-optional-data-on-browser-back-in-angular-ui-router)

**Edit**

Same issue appears for the data option of state, you could look at this [SO Question](http://stackoverflow.com/questions/31542972/how-to-persist-optional-data-on-browser-back-in-angular-ui-router)

**Comment by eddiemonge** [comment]
Im going to close this as it looks like an implementation issue and not a bug. If the SO question deems it is a bug. I can reopen it then.

**Comment by pankajparkar** [comment]
@eddiemonge Could you tell me what is wrong with my current implementation.. As what I think is why the optional parameter of state in not considered while I'm doing browser back.

**Comment by pankajparkar** [comment]
@eddiemonge Would you mind to guide me on this issue in correct direction for implementation..?

**Comment by eddiemonge** [comment]
Im a bit confused by what you are asking honestly

**Comment by pankajparkar** [comment]
@eddiemonge As you said I made wrong implementation.. Can you please go through my description and let me know what is wrong with my current implementation.?
What my concern is `params` is part of the state, why it don't get preserved. On browser back it should repopulate those value.

If  you need further information then let me know.. I'll do it.

**Comment by Okazari** [comment]
I pretty much guess that using browser back button trigger a "new" state change and dont restore the previous state. @eddiemonge  Would be great if you have a workaround (or a solution) about this as it is a pretty much common usecase.

The requested behavior is to "restore" the previous state (especially the controller state) on browser back button (could be great to do with $state.go and ui-sref http://stackoverflow.com/questions/31590223/angular-ui-router-states/31590450#comment51133736_31590450).

Basically it would be to have a better control about moving backward (or forward) in a navigation history. Having the controller values restored with the action (Maybe just not triggering a controller reset). Just imagine that going from a list to a item detail state, then click on the back button (or on a breadcrumb) to go back to the list and having all you controllers values restored (or just not reseted) filters etc...

Is it actually possible with ui-router ? How would you implement this behavior ?

**Comment by pankajparkar** [comment]
@eddiemonge Does the [answer given to my question][1] is good approach or not? Please have some thoughts on it.

[1]: http://stackoverflow.com/a/31620928/2435473

**Comment by christopherthielen** [comment]
Is there a bug? This is an issue tracker for bugs.

**Comment by pankajparkar** [comment]
@christopherthielen Yes..You could go through the conversation. `ui-router` should store `$stateParams` somewhere which are passed optionally in a state, so that on browser back they can get restored from that storage and state will loaded as it was before..

**Comment by nateabele** [comment]
@christopherthielen @eddiemonge Okay guys, I see what's going on.

Basically, there's an inconsistency between how params are handled when they're defined as part of the URL vs. when they're not. I'm actually sort of surprised this has never come up before.

The underlying Push State API lets you pass a state object that you can associate with a URL, which would solve this pretty cleanly, save us some de/serializing overhead on history traversal, and provide higher fidelity for complex param objects, but access to this wasn't exposed in Angular until 1.3.x (see `$location.state()`).

I understand why the behavior is confusing, and while I don't know if I'd go so far as to call it a bug, I think we can reopen this as a feature.

**Comment by phazei** [comment]
Any update with this with angular 1.5/6 and ui-router 1.0?

**Comment by chigia001** [comment]
Another Use Case related to this is when I refresh the page, my expectation is the state params that don't define in URL will also persistent.

Right now it don't persist and revert to default value
